pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -1093,47 +1130,78 @@ func latestPredecessor(
 // picks a random predecessor for the given release version. If we are
 // choosing a predecessor in the same series as the minimum supported
 // version, special care is taken to select a random predecessor that
-// is more recent that the minimum supported version.
+// is more recent that the minimum supported version. Similarly, the
+// same is done for the minimum bootstrap version. minSupported must be
+// set but minBootstrap can be nil.
 func randomPredecessor(
-	rng *rand.Rand, v, minSupported *clusterupgrade.Version,
+	rng *rand.Rand, v, minSupported, minBootstrap *clusterupgrade.Version,
 ) (*clusterupgrade.Version, error) {
 	predecessor, err := release.RandomPredecessor(rng, &v.Version)
 	if err != nil {
 		return nil, err
 	}
 
-	// If the minimum supported version is from a different release
-	// series, we can pick any random patch release.
 	predV := clusterupgrade.MustParseVersion(predecessor)
-	if predV.Series() != minSupported.Series() {
+	// minVersion is either the minimum bootstrap version or the minimum supported version,
+	// depending on the predecessor series. When choosing a predecessor version, we want to
+	// make sure we can run test hooks if possible on that release series. So we need to
```

Comment:
> So we need to pick a version that is at least the minimum supported version or the minimum bootstrapversion, if they are the same series.

The case analysis is getting tricky, let's be precise. The order is minBootstrap <= minSupported < currentVersion, right?

Also, we should probably add another parameter validation to ensure minBootstrap <= minSupported, when both are user-specified.

Reply:
Added, although i moved it to the explanation further down as to not fragment it. Also added a concrete example since I think that makes it clearer whats happening.